Add: tailwindcss view support

benkoshy opened the pull request:
HI @janko 

Please find attached a work in progress PR for Tailwind view support. Some screenshots:

![image](https://user-images.githubusercontent.com/15097447/172290132-750dde21-82f6-4e6a-920c-444cc7a38280.png)

![image](https://user-images.githubusercontent.com/15097447/172290191-bb18415c-5bd9-4b69-8728-39a8794b97f7.png)

### Notes

- the `_login_form_footer` is empty.
- I have only created it for the main pages: e.g. login, reset passwords etc. The following views are still yet to be done:

      /tailwind/add_recovery_codes.html.erb
      /tailwind/recovery_auth.html.erb
      /tailwind/recovery_codes.html.erb
      /tailwind/two_factor_auth.html.erb
      /tailwind/two_factor_disable.html.erb
      /tailwind/two_factor_manage.html.erb
      /tailwind/unlock_account.html.erb
      /tailwind/unlock_account_request.html.erb

- To save you time, I have already created a [tailwind demo app](https://github.com/benkoshy/rodauth-rails-tailwind-demo) which you can spin up in an instant.

- I have retained all the original templates in the `rodauth` directory. You may decide to put the default files, rather, in a `bootstrap` directory.

- The views should be responsive, mobile first, and have dark mode enabled as well. Some of the dark mode views don't work for the full page. I will come back and fix this.

Please do not hesitate to send things back to me to fix etc: tt might be a little more work to delegate, true, but the investment might be worth it.

janko commented:
Looks pretty good 🙂. I noticed that Tailwind templates have some extra content, would it be possible for them to have the same page content as the Bootstrap ones? This way they will be more similar to Bootstrap ones, and they will be fully internationalized when switching locales (e.g. when using rodauth-i18n).

benkoshy commented:
### Questions
> Looks pretty good slightly_smiling_face. I noticed that Tailwind templates have extra text, would it be possible for them to have the same page content as the Bootstrap ones? This way they will be more similar to Bootstrap ones, and they will be fully internationalized when switching locales (e.g. when using rodauth-i18n).

- I don't understand what is meant by standardizing the text: could you pls elaborate?


### Items to Confirm
Please also confirm the following items:

1. Default bootstrap views to be placed in a new bootstrap folder: Yes/No
2. A separate rodauth layout to contain all rodauth views: Yes/No

**Why?** Because if you use the standard `application.html.erb` layout then people might be using all sorts of classes there which would mess up the rodauth styling. Allusion to this was made in #109. This might result in a heavy support/maintenance burden:

```erb
<main class="container mx-auto mt-28 px-5 flex">  <---- Note the classes here            
</main>
```

Others might do this:
```erb
<main> <---- No classes noted here      
</main> 
```
I can change the generators to reflect this.

3. Whether the tests are adequate? Yes/No

janko commented:
> I don't understand what is meant by standardizing the text: could you pls elaborate?

For example, the login form has "Sign In" heading, and "Don't have an account? Sign up here", neither of which are present in the original Bootstrap templates. So, it's extra text/content.

If I use rodauth-i18n and choose a locale with built-in translations (e.g. Portugese), these chunks of custom text will stay in English. This provides a different experienced compared to default Bootstrap templates, which will be fully translated.

I will answer the rest later today 🙂

janko commented:
> Default bootstrap views to be placed in a new bootstrap folder: Yes/No

I'm leaning on the side of keeping the default bootstrap views where they are, to signal that these match Rodauth's [built-in templates](https://github.com/jeremyevans/rodauth/tree/master/templates), making them official in a sense.

> A separate rodauth layout to contain all rodauth views: Yes/No

If we had a special layout, we would also need to make it opt-outable, and I would prefer not to go down that route. I would like to keep the simplicity the Bootstrap views have. Having `<main>` be a horizontal flexbox doesn't seem like a good default to me, I don't expect many people to have that.

> Whether the tests are adequate? Yes/No

They look good to me 👍🏻

benkoshy commented:
@janko  Sir,

> This way they will be more similar to Bootstrap ones, and they will be fully internationalized when switching locales (e.g. when using rodauth-i18n).

I felt that it was important to allow for additional sign-in options (i.e. sign in via Google / Twitter) within the original tailwind template -  this may conflict with the design intent of the library....but if I can make it work with `rodauth-i18n` would you object? In other words, the forms will have some slight differences, but will be fully internationalized. pls confirm, and I can do accordingly as per your directives.

rgds
Ben

janko commented:
> I felt that it was important to allow for additional sign-in options (i.e. sign in via Google / Twitter) within the original tailwind template - this may conflict with the design intent of the library

I don't think it makes sense to do that if Rodauth itself doesn't provide this functionality. I get that it might have been part of a Tailwind UI template, but it's not suitable for rodauth-rails, the default templates should only contain functional elements.

> but if I can make it work with rodauth-i18n would you object? In other words, the forms will have some slight differences, but will be fully internationalized. pls confirm, and I can do accordingly as per your directives.

Having additional translations just for Tailwind templates would be an unnecessary maintenance burden. It couldn't be part of the rodauth-i18n project, because that gem can be used without Rails, so it would likely be missing translations for many languages that get added to rodauth-i18n.

Correct me if I'm wrong, but wasn't the goal of Tailwind templates to have a minimal thing that looks good when you're using Tailwind, just to give you a starting point? In that sense, having templates that look exactly the same as Bootstrap but are styled with Tailwind would fulfill that goal. I'm not saying this is what I want, I just think we should strive for simplicity here.

benkoshy commented:
@janko I had intended for completion today, but am delayed. I try for next Friday. I beg your patience.

kyle-rader commented:
Oh my gosh, I have also been adding Tailwind styles to the rodauth views in my own project this last week. These look great so far, I'd love to adopt them!

benkoshy commented:
@kyle-rader Please be patient with me - I'm still trying to grok the rodauth and rodauth-rails library so i get stuck really easily :'(  ---> they're small things, but big enough to hold me up..........only a few more forms are left to be done, which i will steadily chip away at.

benkoshy commented:
More snips. Compatible in dark mode in the latest absolute latest version of rodauth. working through the last few multifactor forms

![image](https://user-images.githubusercontent.com/15097447/186545702-b6686ec9-8759-4ff5-a186-c63f6eba5c23.png)

![image](https://user-images.githubusercontent.com/15097447/186545758-691e54c0-fd80-4c7c-9e3a-e1b5e08bce8a.png)

benkoshy commented:
@janko If you would like to test out the integration - i have prepared a demo app for you, so you can test it out easily:  
https://github.com/benkoshy/rodauth-rails-tailwind-demo - simply clone and follow the instructions in the readme. 

The only thing that is left to complete is the webauth forms. Which I hope to do soon in a separate PR.

The general goal is to allow people, when creating new apps, to quickly get started, rather than defaulting to other solutions. I confess I'm not the greatest tailwind css expert out there, so there might be a few issues. I have endeavoured to mirror the bootstrap forms where relevant, with minor changes - e.g. centering the forms etc.

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/_login_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+            </div>
+          <% end %>
+        </div>
+        <%= form.submit rodauth.login_button, class: "w-full px-8 py-3 font-semibold rounded-md 
+          w-full flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 
+          dark:bg-emerald-400 dark:text-gray-900" %>
```
I would prefer we consistently put all CSS class names in the same line, no matter how many they are. What do you think?

benkoshy replied:
Fixed: [fc54f92](https://github.com/janko/rodauth-rails/pull/114/commits/fc54f921fd6986ab2b0713602f9e72b54cec5a4c)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/add_recovery_codes.html.erb:
```diff
@@ -0,0 +1,8 @@
+<% content_for :title, rodauth.add_recovery_codes_page_title %>
```
These have been removed in favour of the title instance variable that's now automatically set by rodauth-rails.

benkoshy replied:
fixed with: [101afbe](https://github.com/janko/rodauth-rails/pull/114/commits/101afbeac6751097d6aad074aafe3317b40afe4a)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/change_login.html.erb:
```diff
@@ -0,0 +1,33 @@
+    <h3 class="mb-3 text-3xl font-semibold text-center dark:text-gray-200"> Change Your Login Email Addres </h3>
+    <%= form_with url: rodauth.change_login_path, method: :post, data: { turbo: false } do |form| %>
+      <div class="mx-4 py-1">
+        <!-- The label should probably be named "email"  under rodauth's settings-->
+        <%= form.label "login", "Email", class: "block text-sm text-gray-400 dark:text-gray-100" %>
```
We should use `rodauth.login_label` instead of "Email".

benkoshy replied:
Fixed: [86d1a93](https://github.com/janko/rodauth-rails/pull/114/commits/86d1a936ab82d0e1df32038dbb4f0624af54a5b7)

benkoshy replied:
I'll look out for similar instances too.

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/change_login.html.erb:
```diff
@@ -0,0 +1,33 @@
+<% content_for :title, rodauth.change_login_page_title %>
+<div class="dark:bg-gray-900 flex items-center justify-center h-screen">
+  <div class="max-w-sm">
+    <h3 class="mb-3 text-3xl font-semibold text-center dark:text-gray-200"> Change Your Login Email Addres </h3>
```
With the goal of avoid hardcoded english text, we could use `rodauth.change_login_page_title` here, which says "Change Login" by default. And similar for other forms with titles.

benkoshy replied:
fixed with: [5bb294c286b104c5c2d6f104e5d2a3f2188f4dc0](https://github.com/janko/rodauth-rails/pull/114/commits/5bb294c286b104c5c2d6f104e5d2a3f2188f4dc0)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/change_login.html.erb:
```diff
@@ -0,0 +1,33 @@
+        <!-- The label should probably be named "email"  under rodauth's settings-->
+        <%= form.label "login", "Email", class: "block text-sm text-gray-400 dark:text-gray-100" %>
+        <%= form.email_field rodauth.login_param, value: params[rodauth.login_param], id: "login", autocomplete: "email", required: true, class: class_names("w-full px-3 py-2 border rounded-md border-gray-400 dark:border-gray-700 dark:bg-gray-900 dark:text-gray-100 focus:dark:border-emerald-400 focus:dark:bg-gray-800 invalid:border-red-500 invalid:text-red-600 focus:invalid:border-red-500 focus:i ...
+        <p class="text-xs text-gray-600 dark:text-gray-300 mt-1"> Enter new email address</p>
```
Would the form still look good without this? I would like to avoid using any hardcoded english text, so that the whole form is translated when used with rodauth-i18n.

benkoshy replied:
Fixed here: [2c6b4fe](https://github.com/janko/rodauth-rails/pull/114/commits/2c6b4fef8b8bc9b20ec9eac159b033e237b63a56)

On this point, I'll throw in my two cents:

![image](https://user-images.githubusercontent.com/15097447/192899646-2aa068f2-b715-4bcb-b7f1-39f4ccf23007.png)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/close_account.html.erb:
```diff
@@ -0,0 +1,23 @@
+<% content_for :title, rodauth.close_account_page_title %>
+<div class="dark:bg-gray-900 flex items-center justify-center h-screen">
+  <div class="mx-4 py-4">
+    <h3 class="mb-3 text-3xl font-semibold text-center text-white"> Close Account </h3>
```
I noticed some form titles are `<h2>` and some `<h3>`. Would it make sense to use the same heading for all forms? `<h2>` would seem reasonable to me.

benkoshy replied:
Fixed with: [f0a5307](https://github.com/janko/rodauth-rails/pull/114/commits/f0a5307420273872e47bf8bc9e76e3b1cffd4891)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/close_account.html.erb:
```diff
@@ -0,0 +1,23 @@
+<div class="dark:bg-gray-900 flex items-center justify-center h-screen">
+  <div class="mx-4 py-4">
+    <h3 class="mb-3 text-3xl font-semibold text-center text-white"> Close Account </h3>
+    <div class="flex p-4 mb-4 text-sm text-red-700 bg-red-100 rounded-lg dark:bg-red-200 dark:text-red-800" role="alert">
+      <div>
+        <span class="font-medium">Warning!</span> You're about to close your account!
+      </div>
+    </div>
```
While I see this being useful, I would prefer we don't include it in the default template, since it won't be translated.

benkoshy replied:
> While I see this being useful, I would prefer we don't include it in the default template, since it won't be translated.

Given the translation concerns, I have commented out the text, allowing users to implement what they feel is best, because I thought the gravity of some type of warning necessary. If this is unsuitable, please LMK and I will immediate omit it altogether: [76f8fae](https://github.com/janko/rodauth-rails/pull/114/commits/76f8fae42b65d3a508f7225cea26caaf995e2208)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/logout.html.erb:
```diff
@@ -0,0 +1,14 @@
+      <%= form.submit rodauth.logout_button, class: "px-8 py-2 font-semibold rounded-md flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 dark:bg-emerald-400 dark:text-gray-900" %>
+    </div>
+  </div>
+<% end %>
```
All files should have a trailing newline, you'll see GitHub warning you in the diff when it's missing 😉

benkoshy replied:
Fixed with [1a8c11f](https://github.com/janko/rodauth-rails/pull/114/commits/1a8c11f5fb367862b7b2621f0f446fc0ec614eb5)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/otp_setup.html.erb:
```diff
@@ -0,0 +1,52 @@
+				<div class="w-full">
+					<div>
+						<p class="dark:text-gray-100 ">
+							<b><%= rodauth.otp_secret_label %>: </b>
```
MDN [advises against](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b) using the `<b>` tag, and instead recommends using `<strong>` or styling with CSS.

benkoshy replied:
Fixed with: [2a2f6ba](https://github.com/janko/rodauth-rails/pull/114/commits/2a2f6ba8edf0b16a3e2842dc793725f8f8e47c0b)

janko commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/reset_password_request.html.erb:
```diff
@@ -0,0 +1,25 @@
+        <%= form.hidden_field rodauth.login_param, value: params[rodauth.login_param] %>
+      <% else %>
+        <div class="py-4">
+          <%#= form.label "login", rodauth.login_label, class: "block text-sm text-gray-600" %>
+          <!-- The label is unecessary,  given the email placeholder below: -->
```
Should we remove these comments?

benkoshy replied:
[fixed](https://github.com/janko/rodauth-rails/pull/114/commits/4bede9b975ca3fd7e812df14d56a709942516d94)

janko commented on lib/generators/rodauth/views_generator.rb:
```diff
@@ -96,11 +100,23 @@ def configuration_name
         def form_helpers_compatibility(content)
           content
-            .gsub(/form_with url: (.+) do \|form\|/, 'form_tag \1 do')
-            .gsub(/form\.(label|submit)/, '\1_tag')
-            .gsub(/form\.(email|password|text|telephone|hidden)_field (\S+), value:/, '\1_field_tag \2,')
-            .gsub(/form\.radio_button (\S+), (\S+),/, 'radio_button_tag \1, \2, false,')
-            .gsub(/form\.check_box (\S+), (.+) /, 'check_box_tag \1, "t", false, \2 ')
+          .gsub(/form_with url: (.+) do \|form\|/, 'form_tag \1 do')
+          .gsub(/form\.(label|submit)/, '\1_tag')
+          .gsub(/form\.(email|password|text|telephone|hidden)_field (\S+), value:/, '\1_field_tag \2,')
+          .gsub(/form\.radio_button (\S+), (\S+),/, 'radio_button_tag \1, \2, false,')
+          .gsub(/form\.check_box (\S+), (.+) /, 'check_box_tag \1, "t", false, \2 ')
```
Please undo the indentation change.

benkoshy replied:
[fixed](https://github.com/janko/rodauth-rails/pull/114/commits/ee23646bcdaa0641c17f0bf189001413b14c895b). 

The culprit was a 'beautify ruby' gem I was using that automatically formats on save. I'll watch for similar in case there is a reversion.

janko commented on lib/generators/rodauth/views_generator.rb:
```diff
@@ -14,6 +14,10 @@ class ViewsGenerator < ::Rails::Generators::Base
           desc: "Generates views for all Rodauth features",
           default: false
 
+        class_option :css, type: :string,
+          desc: "Generates tailwind css specific views",
+          default: "bootstrap"
```
I was thinking it would be nice if the `:css` option defaults to `tailwind` when `tailwindcss-rails` gem is used. I like how in this case Rails' built-in scaffold automatically generates Tailwind-friendly templates, and would like to have the same behaviour for rodauth-rails 🙂

benkoshy replied:
I can do the same here.

Q: Would you still want to retain the `class_option`?

I would throw in my two cents on the above question: 

- going forward on rails +7 apps, the canonical approach would probably be to use the [`cssbundling-rails`](https://github.com/rails/cssbundling-rails) gem. i.e. via node, rather than the `tailwindcss-rails` gem.
- Would there be other styles you would want rodauth-rails to perhaps support: e.g. `semantic-ui` etc.

janko replied:
Yes, I would prefer to keep the class option, to enable generating Tailwind templates for Rails apps using Tailwind without `tailwindcss-rails` (e.g. via `cssbundling-rails` that you mentioned).

benkoshy replied:
[3cc210b](https://github.com/janko/rodauth-rails/pull/114/commits/3cc210b8b8a4c1227ce98f732d8742a9438dce9d)

janko commented on test/generators/views_generator_test.rb:
```diff
@@ -62,6 +62,37 @@ class ViewsGeneratorTest < Rails::Generators::TestCase
+		run_generator %w[--css=tailwind logout]
+
+	  assert_file "app/views/rodauth/logout.html.erb", <<~ERB.strip()
+<% content_for :title, rodauth.logout_page_title %>
+<div class="flex justify-center">
+  <%= form_with url: rodauth.logout_path, method: :post, data: { turbo: false } do |form| %>
+    <% if rodauth.features.include?(:active_sessions) %>
+      <div class="py-3">
+        <%= form.check_box rodauth.global_logout_param, id: "global-logout", class: "appearance-none checked:bg-blue-500" %>
+        <%= form.label "global-logout", rodauth.global_logout_label, class: "inline-block text-gray-800" %>
+      </div>
+    <% end %>
+    <div class="py-3">
+      <%= form.submit rodauth.logout_button, class: "px-8 py-2 font-semibold rounded-md flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 dark:bg-emerald-400 dark:text-gray-900" %>
+    </div>
+  </div>
+<% end %>
```
I think it's enough to just match a Tailwind-specific utility class instead of the whole template, because this way we'll need to copy-paste the template each time we make a change there. And we can include this assertion in the first test, I don't think we need two.

benkoshy replied:
I wasn't sure how to test for a utility specific class from the template. So I opted for this approach instead: [3cc210b](https://github.com/janko/rodauth-rails/pull/114/commits/3cc210b8b8a4c1227ce98f732d8742a9438dce9d) - basically extracting the whole template into a private method. I agree it's less than ideal. I can run a simple integration test to test against the utility specific class.

Do not hesitate to revert back to me if required.

janko commented on test/generators/views_generator_test.rb:
```diff
@@ -62,6 +62,37 @@ class ViewsGeneratorTest < Rails::Generators::TestCase
+		ERB
+	
+
+  end if ActionView.version >= Gem::Version.new("5.1")
```
Why don't the tests pass on older Rails versions?

benkoshy replied:
Fixed: [3cc210b](https://github.com/janko/rodauth-rails/pull/114/commits/3cc210b8b8a4c1227ce98f732d8742a9438dce9d) - no idea why i added that.

benkoshy commented on lib/generators/rodauth/templates/INSTRUCTIONS:
```diff
@@ -33,8 +33,24 @@ Depending on your application's configuration some manual setup may be required:
+              require('@tailwindcss/forms'),
+          ],
+          darkMode: 'class' // one way of enabling dark mode
+       }
```
```diff
-
-       If you are using tailwind, please ensure you have: 
-
-       -(i) the tailwind forms plugin installed (e.g. `yarn add @tailwindcss/forms`)
-
-       -(ii) dark mode enabled (in some way) i.e. see: https://tailwindcss.com/docs/dark-mode#toggling-dark-mode-manually, should you want it:
-
-       module.exports = {  
-          plugins: [      
-              require('@tailwindcss/forms'),
-          ],
-          darkMode: 'class' // one way of enabling dark mode
-       }
+          # If using tailwind, ensure you have tailwind forms plugin installed (e.g. `yarn add @tailwindcss/forms`).
```
[Commit: a0108ff](https://github.com/janko/rodauth-rails/pull/114/commits/a0108ffc1d89476ff9aca25169d4fb302130bcfb)

Please let me know if this is inadequate.

janko replied:
Including the specific command (`yarn add @tailwindcss/forms`) suggests that this is the *only* thing needed to enable the plugin, but the developer also needs to add it to the list of plugins in `tailwind.config.js`. That's why I would prefer a more vague wording like what I proposed, and rely on the developer to know the logistics.

benkoshy commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/_login_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+            <div class="space-y-2">
+              <%= form.label "login", rodauth.login_label, class: "block text-sm text-gray-400" %>
+              <%= form.email_field rodauth.login_param, value: params[rodauth.login_param], id: "login", autocomplete: "email", required: true, class: class_names("w-full px-3 py-2 border rounded-md border-gray-400 dark:border-gray-700 dark:bg-gray-900 dark:text-gray-100 focus:dark:bg-gray-800 focus:dark:border-emerald-400 invalid:border-red-500 invalid:text-red-600 focus:invalid:border-red-500 f ...
+              <!-- up to here -->
```
oops. [fixed](https://github.com/janko/rodauth-rails/pull/114/commits/a2ff69beb3b11472d5cc85abca4814def99b9ee3)

benkoshy commented on lib/generators/rodauth/templates/INSTRUCTIONS:
```diff
@@ -33,8 +33,24 @@ Depending on your application's configuration some manual setup may be required:
+       
+
 
-     * Not required *
```
[fixed](https://github.com/janko/rodauth-rails/pull/114/commits/e15b9cf94c62ee52410c2ddd68524e7252ce09c9)

benkoshy replied:
oops. fixed with [e15b9cf](https://github.com/janko/rodauth-rails/pull/114/commits/e15b9cf94c62ee52410c2ddd68524e7252ce09c9)

benkoshy commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/_login_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+            <div class="space-y-2">
+              <div class="flex justify-between">
+                <%= form.label "password", rodauth.password_label, class: "block text-sm text-gray-400" %>
+                <%= link_to("Forgot password?", rodauth.reset_password_request_path, class: "text-xs text-blue-600 hover:text-blue-700 hover:underline dark:text-gray-200") %>
```
fixed with [865ac51](https://github.com/janko/rodauth-rails/pull/114/commits/865ac514b5decee06b299bcd7dd7396ff4cf6103)

benkoshy commented:
> ...I will try to test out the templates later this week.

Consider holding off till I push through the above changes - you may save time if you do so.

benkoshy commented on lib/generators/rodauth/templates/app/views/rodauth/tailwind/_login_form.html.erb:
```diff
@@ -0,0 +1,47 @@
+          <% if rodauth.skip_login_field_on_login? %>
+            <div class="space-y-2">
+              <%= form.label "login", rodauth.login_label, class: "block text-sm text-gray-400 dark:text-gray-200" %>
+              <%= form.email_field rodauth.login_param, value: params[rodauth.login_param], id: "login", readonly: true, class: class_names("w-full px-3 py-2 border rounded-md border-gray-400 dark:border-gray-700 dark:bg-gray-900 dark:text-gray-100 focus:dark:border-emerald-400 focus:dark:bg-gray-800 invalid:border-red-500 invalid:text-red-600 focus:invalid:border-red-500 focus:invalid:ring-red-5 ...
```
fixed with: [0d2e614](https://github.com/janko/rodauth-rails/pull/114/commits/0d2e6146c9b59d0a32724f373eeddca1c7e543c8)

benkoshy commented:
@janko  Feel free to try out it out on the demo app here: https://github.com/benkoshy/rodauth-rails-tailwind-demo - instructions contained therein.

Also I have hard-coded to a dark-mode color. This may not be the best outcome. Perhaps it is better to specify dark mode for the elements within the form, rather than the dark mode itself. This will be patent on testing the demo app above.

Feel free to push back onto me anything that needs reworking.

janko commented:
When I look at the login page, I see numerous issues:

1. input fields are red by default (which is bad UX, we should probably rely just on server-side validation errors)
1. when I focus a valid field, the blue focus outline mixes with the green border
1. labels are too light, the contrast doesn't even meet accessibility standards (why not black labels?)
1. that box shadow looks out of place (maybe it shouldn't be there?)
1. the inline "Forgot Password?" is not vertically aligned with "Login" label (do we really need it, given that we already have one at the bottom?)
1. input fields seem too narrow (a longer email/password won't fit)

<img width="300" alt="Screenshot 2022-10-04 at 10 46 56" src="https://user-images.githubusercontent.com/795488/193776211-0cc3faf5-eb1f-44d3-9c46-8b453b269781.png">

In the dark mode, I think it would be better if forms just inherited the background color of the container, rather than having their own. That's how Bootstrap templates behave, and it seems like a more sensible default.

When I look at the create account form, the color of labels and spacing between fields is totally different from the login form. The form layouts should all be consistent.

<img width="250" alt="Screenshot 2022-10-04 at 11 01 34" src="https://user-images.githubusercontent.com/795488/193779448-bbf146ab-f88a-46b1-818e-7468e77135ad.png">

Also, when I cause validation errors in the create account form, for some reason the input fields expand to double the length 🤷🏻‍♂️. Also, the validation error message color is different from the border of the invalid field, and I think it would look better if it was the same.

<img width="350" alt="Screenshot 2022-10-04 at 11 03 21" src="https://user-images.githubusercontent.com/795488/193780131-c66d03d4-aa0e-4f6f-8061-374750a133f9.png">

In the TOTP setup form, the input fields are way too narrow, the "Authentication Code" label even spans multiple lines, not to mention that the button text is cut off.

<img width="693" alt="Screenshot 2022-10-04 at 11 16 58" src="https://user-images.githubusercontent.com/795488/193782719-3f86d1aa-638a-4e83-af7a-29c07740d2cf.png">

I see many more things:

* in the reset password request form, the login field has a "login" placeholder (which is inconsistent with other forms which don't have any placeholders)
* in the verify account resend form, the login field doesn't have a label (which is inconsistent with other forms)
* the change password/login forms display at the bottom of the page for some reason

Also, there seem to be some merge conflicts with the test file for the views generator.

benkoshy commented:
I appreciate the feedback.

I will incorporate the above suggestions.

I propose then to move forward incrementally: 

1. let me get one form absolutely correct: e.g. the login page.
2. And then apply the same model to the rest, moving incrementally.

benkoshy commented:
@janko I beg your patience. I have had some constraints on my end. I wish to see this through to completion. Fingers crossed by next week.

benkoshy commented:
Hi @janko  - I have updated just one form - the login form in light of your comments - if you find this adequate, I will update the same across all the forms. If you have a spare 5 min - i would not spend any more than that -  and could give it a review, I would appreciate that

>     input fields are red by default (which is bad UX, we should probably rely just on server-side validation errors)
>     when I focus a valid field, the blue focus outline mixes with the green border
>     labels are too light, the contrast doesn't even meet accessibility standards (why not black labels?)
>     that box shadow looks out of place (maybe it shouldn't be there?)
>     the inline "Forgot Password?" is not vertically aligned with "Login" label (do we really need it, given that we already have one at the bottom?)
>     input fields seem too narrow (a longer email/password won't fit)
> 

Link: https://github.com/benkoshy/rodauth-rails-tailwind-demo
`./bin/dev`

Route: http://localhost:3000/login

(I will also remove the hard coding of the particular dark-mode color)

Please do not hesitate to send back any problems. thank you for your patience.

Ben

janko commented:
I ended up merging the current work, and fixed various issues I encountered, added some improvements, and simplified markup where it made sense. Thanks for all your hard work in providing a great base for this feature 🙏🏻

benkoshy commented:
> I ended up merging the current work, and fixed various issues I encountered, added some improvements, and simplified markup where it made sense. Thanks for all your hard work in providing a great base for this feature 🙏🏻

That would have taken some time! Your contributions have been immense - I regret i could not have alleviated your load more. 😔
